xplique/attributions/image_object_detector_explainer.py

```diff
@@ -0,0 +1,280 @@
+        # return the intersection over union value
+        return iou
+
+
+class IObjectFormater:
+    """
+    Generic class to format the model prediction
+    """
+    def format_objects(self, predictions) -> Iterable[Tuple[tf.Tensor, tf.Tensor, tf.Tensor]]:
+        """
+        Format the model prediction of a given image to have the prediction of the following format:
+        objects, proba_detection, one_hots_classifications
+
+        Parameters
+        ----------
+        predictions
+            prediction of the model of a given image
+
+        Returns
+        -------
+        object
+            bounding box or mask component of the prediction
+        proba
+            existence probability component of the prediction
+        classification
+            classification component of the prediction
+        """
+        raise NotImplementedError()
+
+
+class ImageObjectDetectorScoreCalculator:
+    """
+    Class to compute batch score
+    """
+    def __init__(self, object_formater: IObjectFormater, iou_calculator: IouCalculator):
+        self.object_formater = object_formater
+        self.iou_calculator = iou_calculator
+
+    def batch_score(self, model, inputs, targets, batch_size):
+        """
+        Compute batch score
+
+        Parameters
+        ----------
+        model
+            the model used for the object detection
+        inputs
+            Input samples to be explained.
+        targets
+            model outputs
+        batch_size
+            number of masked samples to explain at once, if None process all at once.
+        """
+        return inference_batching(self.score, model, inputs, targets, batch_size)
+
+    def score(self, model, inp, object_ref) -> tf.Tensor:
+        """
+        Compute the matching score between prediction and a given object
+
+        Parameters
+        ----------
+        model
+            the model used for the object detection
+        inp
+            the batched image
+        object_ref
+            the object target to compare with the prediction of the model
+
+        Returns
+        -------
+        score
+            for each image, the matching score between the object of reference and
+            the prediction of the model
+        """
+        objects = model(inp)
+        score_values = []
+        for obj, obj_ref in zip(objects, object_ref):
+            if obj is None:
+                score_values.append(tf.constant(0.0, dtype=inp.dtype))
+            else:
+                current_boxes, proba_detection, classification = \
+                            self.object_formater.format_objects(obj)
+
+                if len(tf.shape(obj_ref)) == 1:
+                    obj_ref = tf.expand_dims(obj_ref, axis=0)
+
+                obj_ref = self.object_formater.format_objects(obj_ref)
+
+                scores = []
+                size = tf.shape(current_boxes)[0]
+                for boxes_ref, proba_ref, class_ref in zip(*obj_ref):

```

Comment:
Not an obligation but It could be cool to have some comments here to understand what' going on. I will try to adress that

Reply:
Cf paper. The algorithm implement the process described in the article about D-Rise.